### PR TITLE
Add emote picker widget

### PR DESCRIPTION
**Pull request by Nogesma**
This PR adds an emote picker widget, accessible with the (Alt + e) key combination when the chat input widget is open. (key combination can change).

They was a few challenges with introducing this, so this is unfortunately quite a large diff. 

I had to refactor a lot of the graphics protocol code as decoding images, especially animated webp is quite slow.
There is now a separate thread used for decoding images (only used for decoding emotes requested by the emote picker widget for now).
Although the decoding happens in another thread, it is still the main thread that handles sending graphics protocol commands, as it involves writing to stdout, and I noticed some freezes when I tried locking stdout from the decoding thread.

I also switched to using the webp-animation (which uses libwebp) crate for decoding webp, as it is on average twice as fast as the current image.rs implementation. Image.rs has a WIP rework of their webp implementation [here](https://github.com/image-rs/image-webp), I'll probably look into switching back once they are done, and if the performance is comparable.

For the emote picker widget I tried using the fuzzy finder like in the channel switcher widget, but it was a bit slow, especially with some channels having a few thousand emotes. I switched to using the memchr crate.

Another issue I encountered was that kitty can become quite slow after loading a lot of images/frames (starting at a few thousands on my system), and with some emotes having a few hundred frames, it can become quite slow scrolling through the emotes. I might look into manually keeping track of the number of images we sent to kitty and cleaning up the ones we haven't been using in a while, if this issue becomes prevalent.

I also fixed a bug where emotes were not actually being cleared, because kitty keeps two separate image list for normal/alternate screen and I was clearing the emotes after we left the alternate screen.

I probably need to clean up a bit the widget code but otherwise this is pretty much done.

Let me know if you find any issues or changes you need me to make.

### CODE REVIEW

**Comment by Nogesma** [comment]
Okay I finished cleaning up the code a bit, this is ready for review.

**Comment by Xithrius** [comment]
I shall take a stare in the coming days

**Comment by Nogesma** [comment]
Thanks, if you notice some lag when scrolling through the emote list, a fix is available on the [kitty vt branch](https://github.com/kovidgoyal/kitty/tree/vt).

**Review by Xithrius — CHANGES_REQUESTED** [review]
Looking pretty good so far. Not seeing anything that's far out of the ordinary for component implementation. Tested on a release build, and I'm not getting any noticeable lag.

I think the only thing that should probably change is making the search input box case insensitive.

Very solid work. Thank you very much for all that you've done!

**Comment by Nogesma** [comment]
The reason I was not using the fuzzy matcher, was that it was a bit slow to match against the sometime thousands of emotes.
I just pushed a commit using it so that you can test yourself, but I can notice a bit of lag when typing fast in the emote search box, with channels like thebausffs that have tons of emotes.
If it's acceptable for you feel free to merge, otherwise I can maybe look into moving the whole draw call into something async, to at least have a responsive input.

File: src/ui/components/emote_picker.rs [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+        for (name, (filename, zero_width)) in self.emotes.emotes.borrow().iter() {
+            if items.len() >= max_len {
+                break;
+            }
+
+            // Skip emotes that do not contain the current input, if it is not empty.
+            let Some(pos) = finder
+                .as_ref()
+                .map_or_else(|| Some(0), |f| f.find(name.as_bytes()))
```
Comment (Xithrius) [inline src/ui/components/emote_picker.rs]: I find that this works fine with the `memmem::Finder` implementation of the search widget. Thoughts?

```suggestion
                .map_or_else(|| Some(0), |f| f.find(name.to_lowercase().as_bytes()))
```

Comment (Xithrius) [reply src/ui/components/emote_picker.rs]: image

Comment (Nogesma) [reply src/ui/components/emote_picker.rs]: Yes, I did not think about that. I'll implement it tomorrow, thanks.

**Comment by Nogesma** [comment]
Should be good now

**Review by Xithrius — APPROVED** [review]
All is well, all is good. Thank you for the PR!

**Comment by Xithrius** [comment]
Exists in release https://github.com/Xithrius/twitch-tui/releases/tag/v2.6.4

**Comment by Nogesma** [comment]
Nice thanks a lot!

**Pull request merged**